### PR TITLE
enable warnings on cuda synchronization 

### PR DESCRIPTION
**Pull request by ngimel**
This creates `torch.cuda.set_warn_on_synchronization()` function that would warn or error when synchronizing operation is performed. We could wrap it in a context manager for ease of use, but it would be a lie, because it sets global, and not thread-local state. Since it's intended for debugging, maybe that's ok though.  
As all `torch.cuda.*` functions, it's going through CPython, not pybind, so the argument is converted to long before being passed to c10 function. I'll make python argument a python enum class, but without pybind it'll still have to go thourgh long conversion. 

For a test script
```
import torch
torch.cuda.set_sync_debug_mode(1)
x=torch.randn(10, device="cuda")
x.nonzero()
y=torch.randn((), device="cuda")

if y:
    print("something")
torch.multinomial(x.abs(), 10, replacement=False)
torch.randperm(20000, device="cuda")
ind = torch.randint(10, (3,), device="cuda")
mask = torch.randint(2, (10,), device="cuda", dtype=torch.bool)
val = torch.randn((), device="cuda")
x[mask]=1.
x[mask] = val
torch.cuda.synchronize()
```
the output is 
```
/../playground/sync_warn_test.py:4: UserWarning: called a synchronizing operation (Triggered internally at  ../c10/cuda/CUDAFunctions.cpp:145.)
  x.nonzero()
/../playground/sync_warn_test.py:7: UserWarning: called a synchronizing operation (Triggered internally at  ../c10/cuda/CUDAFunctions.cpp:145.)
  if y:
something
/../playground/sync_warn_test.py:9: UserWarning: called a synchronizing operation (Triggered internally at  ../c10/cuda/CUDAFunctions.cpp:145.)
  torch.multinomial(x.abs(), 10, replacement=False)
/../playground/sync_warn_test.py:15: UserWarning: called a synchronizing operation (Triggered internally at  ../c10/cuda/CUDAFunctions.cpp:145.)
  x[mask] = val
```

### CODE REVIEW

**Comment by ezyang** [comment]
Adding new state to c10 seems fine. Something similar would have been helpful in https://github.com/pytorch/pytorch/issues/62031 

Global is fine and consistent with deterministic.

File: c10/cuda/CUDAFunctions.cpp [hunk 1]
```diff
@@ -138,5 +138,14 @@ void device_synchronize() {
   C10_CUDA_CHECK(cudaDeviceSynchronize());
 }
 
+void warn_or_error_on_sync() {
+  if (warning_state().get_sync_warning_level() == SyncWarningLevel::L_ERROR) {
+    TORCH_CHECK(false, "called a synchronizing operation");
```
Comment (mruberry) [inline c10/cuda/CUDAFunctions.cpp:143]: CUDA operation

File: c10/cuda/CUDAFunctions.cpp [hunk 2]
```diff
@@ -138,5 +138,14 @@ void device_synchronize() {
   C10_CUDA_CHECK(cudaDeviceSynchronize());
 }
 
+void warn_or_error_on_sync() {
+  if (warning_state().get_sync_warning_level() == SyncWarningLevel::L_ERROR) {
```
Comment (mruberry) [inline c10/cuda/CUDAFunctions.cpp:142]: make name consistent with Python

File: c10/cuda/CUDAFunctions.h [hunk 3]
```diff
@@ -35,6 +35,28 @@ C10_CUDA_API void set_device(DeviceIndex device);
 
 C10_CUDA_API void device_synchronize();
 
+C10_CUDA_API void warn_or_error_on_sync();
+
+enum class SyncWarningLevel { L_DISABLED = 0, L_WARN, L_ERROR };
+
+class WarningState {
```
Comment (mruberry) [inline c10/cuda/CUDAFunctions.h:42]: Add comment describing this class and intended future use

File: c10/cuda/CUDAFunctions.cpp [hunk 4]
```diff
@@ -138,5 +138,14 @@ void device_synchronize() {
   C10_CUDA_CHECK(cudaDeviceSynchronize());
 }
 
+void warn_or_error_on_sync() {
```
Comment (mruberry) [inline c10/cuda/CUDAFunctions.cpp:141]: Add comment re when and how this is meant to be used

File: torch/csrc/cuda/Module.cpp [hunk 5]
```diff
@@ -439,6 +439,31 @@ PyObject * THCPModule_memorySnapshot(PyObject *_unused, PyObject *noargs)
   END_HANDLE_TH_ERRORS
 }
 
+PyObject * THCPModule_cudaSetSyncDebugMode(PyObject * _unused, PyObject * arg){
+  HANDLE_TH_ERRORS
+  THPUtils_assert(THPUtils_checkLong(arg), "invalid argument to set_warn_on_synchronization");
```
Comment (mruberry) [inline torch/csrc/cuda/Module.cpp:444]: update name

File: torch/cuda/__init__.py [hunk 6]
```diff
@@ -492,6 +492,37 @@ def current_blas_handle():
     _lazy_init()
     return torch._C._cuda_getCurrentBlasHandle()
 
+def set_sync_debug_mode(debug_mode: Union[int, str]) -> None:
+    r"""Sets the debug mode for cuda synchronizing operations.
```
Comment (mruberry) [inline torch/cuda/__init__.py:496]: nit: "CUDA"

File: torch/cuda/__init__.py [hunk 7]
```diff
@@ -492,6 +492,37 @@ def current_blas_handle():
...
+
+    Args:
+        debug_mode(str or int): if "default" or 0, don't error or warn on synchronizing operations,
+            if "warn" or 1, warn on synchronizing operations, if "error" or 2, error out synchronizing operations.
+
+    Note:
```
Comment (mruberry) [inline torch/cuda/__init__.py:502]: Upgrade to warning

File: torch/csrc/cuda/Module.cpp (hunk 5 above)
Comment (mruberry) [reply torch/csrc/cuda/Module.cpp:444]: nit: TORCH_WARN_ONCE when this is called that the feature is a prototype and may not detect all CUDA syncs (yet)

File: torch/testing/_internal/common_utils.py [hunk 8]
```diff
@@ -500,6 +500,17 @@ def __enter__(self):
     def __exit__(self, exception_type, exception_value, traceback):
         torch.use_deterministic_algorithms(self.deterministic_restore)
 
+class CudaSyncGuard:
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:503]: Add a comment for the guard (incl. briefly why it's not exposed)

**Review by mruberry — APPROVED** [review]
Cool! 

Inline comments are from joint review.

**Comment by ezyang** [comment]
deferring to mruberry on this. Thanks for doing this!

**Pull request closed without merging**